### ISSUE TITLE
ESL 读书笔记 (chapter 2)

### ISSUE THREAD

**Issue by yujingma45** [yujingma45/Blog#8]
由于某些契机，已经毕业三年的我又重新捡起来The Elements of Statistical Learning 这本统计学经典。通过两周阅读，才发现自己之前漏掉了很多这本书的精华，甚至字里行间的都包含着很多有用的信息，简直和读一本阿婆的侦探小说一样，有趣又烧脑。这篇笔记会从两个方面深入：行文脉络 （作者到底讲了什么） 和 一些个人知识拾遗。

## 整体思路

第二章算是对supervised learning 的一次整体的讲解，从各个方面都概括了下，形成了完整的思路。本章主要是从最简单的两种方法：linear regression和knn 开始入手，介绍了supervised learning的两种思想：generalize和local； 然后讲了如何优化问题，最后介绍两类思路之间的不同和各自的优缺点，以及过度到smoothing complexity parameters, 如何做到bias vs. variance trade-off.

下面是具体的内容：
### 2.1–2.2 讲了problem type 和variable type；  
首先讲了不同的variable type：categorical variable (qualitative variable, ordered variable) 以及numeric variable (quantitative)；categorical variable中包含两个及两个class 以上的categorical variabel，这里特别点出了dummy variable的具体意义，是在多个classes 之下，如果每个class 之前没有优先顺序，比如收入高中低可以用3，2，1； 但是顾客的种类却不能分为1，2，3， 因为这些用户之间是没有先后优先顺序的。所以对于这种n个类别，但又没有先后顺序的variable 来讲，用 n-1 个类别来表示即可 （one hot encoding 大概也是这么个思路），这就是dummy variable。
其次强调了supervised learning 中根据output 不同，分为regression 和 classification，其中作者特别强调了 classification中的input y一般都是outcome（比如head/tail），但最终output 会是一件事发生的概率，这里我们是根据已有发生的结果（outcome）来预测发生的概率。
### 2.3 开始引入linear regression 和knn； 
linear regression 来讲，其本身是一个有参数而且比较generalize 的方法；knn则是non-parameter 而且比较local的方法。这里值得注意的是，两种方法对同一个问题的不同解答，体现了两种模型看问题的角度：linear regression是从全局上看问题 ，这样的全局一般需要很多的假设来支持才能形成一个模型；而knn 是从局部看问题，所以假设几乎不会有，但是是不是真的局部，是不是能以偏概全又是另一个问题。总体上，linear regression 用的是global linear function来解释问题，而knn 用的是local constant 来解答。
### 2.4讲了如何进行统计决策
基本思路：定义loss function →通过选择不同的参数，来让loss function最小 （损失最少的信息）-> 最终选定模型
首先对问题优化需要一个基本的loss function，用来计算预测值和真实值之间的距离，是衡量你有多少问题没有被解答的方式。为了minimize loss function，就不得不说到optimization（即优化的方法， 让我们尽可能找到一组parameter 来解答尽可能多的问题。就如豆瓣大刀的读书笔记中所讲的：在统计学中，损失函数有很多种，而损失函数归根到底，也就是在某种分布下的期望。
书中用了最常见的一种loss function - MSE作为例子, 为了让这个loss function的损失最小，我们从这个目的出发选择最小化mse, 得到最优解应该是conditional mean。在优化mse 上， linear regression和 knn 各自都在用不同的方法接近conditional mean：linear regression用的是最简单最常见的最小二乘法（但由于不是所有function都可导，所以最小二乘法本身的局限性还是不小的）； 但是knn 则是用neighborhood的mean 来作为结果，然后进行预测，如果样本数量越大，在每个数据点都uniform distributed 的前提下（各个数据点均匀分布），N 越大，和我们要预测的点相同的邻居就会越多，然后我们就越有可能接近conditional mean。
这里还提到另一种loss function- mean absolute error (L1范数的损失函数)， 而minimize mean absolute error 的最优解是conditional median，这个解法最终给出的预测更加robust， （因为median 不受到极值影响）但也由于这个解法的微分不连续（嗯，还是求导问题），很难进行优化所以不常用（关于为什么不连续就不好优化，可以类比成有个地图没有一个完整的线路，中间断断续续，所以你不知道怎么走才好）。
这里还简单提到了classfication 问题的优化方法，主要是hinge loss 和0–1 loss。这里不得不提，个人认为hinge loss 更加合理，因为hinge loss 对做错的分类有不同的对待方式，它对错误有一个容忍范围（margin）然后对错的离谱的点进行更有力度的惩罚，所以它不仅仅关心错不错，还关心有多错。回到loss function上来，之前的regression problem 我们知道要minimize mse 就要找conditional mean， 而在categorical问题上如果要minimize0–1 loss function最合理的方法就是把一个样本分到它最优可能是的类别里，也就是bayes classifier， 而base error 就是bayes classifier 的error rate。
### 2.5 讲了knn 的局限 - the curse of dimensionality
由于knn 是local based lazy learner ，基本不学习而且用例子来帮助预测。 所以最大的问题是，当local 不再是local 我们就无法再使用knn。在高维度时，主要有一下几个原因，1. 同样的半径在低维和高维画出neighborhood包含的信息越来越少; 2. 邻居大多分布在分界线的边缘，大多的邻居距离在一半以上（n=500);3.邻居越来越稀疏； 总体上邻居少而且越来越不同了
### 2.6讲了模型，supervised learning和function approximation 之间的联系
首先特别强调了y 和x 之间存在的两种关系会对model本身选择造成影响：
- Statistical relationship: Conditional probability P(y|x) depends on x only through the conditional mean. 即y本身包含 signal + noise，y 与x 本身的关系并非一定的（not deterministic） 有其他的feature 来影响y，比如说测量误差之类。linear regression中大多也是假定这种关系。这种关系经常用additive error model来解释； 并且假设无法被x解释的部分会独立于x，而x只需要解释在给定x这种条件下的y即可。注意我们在这里假设error term iid。
- Deterministic relationship: 也就是y和x 的关系本身不包含其他的信息，这里的function是deterministic 的。可以认为x本身就包含了randomness，这一类关系主要由error-based model 解决。

这两种relationship解释比较好的是https://newonlinecourses.science.psu.edu/stat414/node/276/

在选择了建立什么样的model之后，本章就进一步阐述了supervised learning的意义，就是在每次比较真值和预测值计算预测误差后，不断根据这些误差来modify model。
那么在选定了model，和具体的做法，我们要怎么不断的让预测值尽可能接近output？这就需要我们尽可能找到一组系数来接近真值， 除了上文中通过minimize error 的方法（最小二乘法），我们还可以找到一组系数让现有看到的事实/现象出现的概率最大化 (mle, 极大似然法），mle 在大多数不可导的情况下非常管用，比如logistic regression

### 2.7 讲了local/generalize 这两种模型结构的影响
除了高纬度的灾难，在低纬度分割过细腻，也会让模型变得更加复杂，而无法apply到这一类问题上，造成overfitting，这里特别强调了越local，越趋近于用常数来作为解答，而越general则会倾向于运用带有很多假设和限制的方程来解答，而model本身的结果对限制的选择是很敏感的。
这里带回到之前我们把local/general 当作局部和全局角度解决问题的类比中去，实际上我们会发现当要宏观的解决一个问题时，我们会做很多假设来支持之后的解决方案，而每一个假设都影响着最后的结果；而局部的解决问题，我们可能就比较单一的去解决， 不会依赖于过多的假设，但也无法把解决方法普及到这一类问题上。
### 2.8 讲了如何在local和general之间进行调节
如果太在乎局部的表现，我们会倾向于复杂的模型，但是另一方面为了让解决方法可以推广，我们有需要smoothing parameter 避免我们看问题的时候过于局限，进而达到**control the effective size**of the local neighborhood
这里提到了几种常用的smooth方法：
- penalty function， 在loss function 加入一个对parameter 惩罚项，从而达到让函数更加简化的目的。 这里可以想到的有lasso，ridge， 和decision tree pruning method
- kernel methods，把问题投射到更高维度来解决。 有些在低维度看起来非常复杂的问题，会在投射到高维度后变得非常简单。 这里可以看作是对点和目标点之前的关系中加了权重，相当于是将最近邻方法中的求均值变成了求加权平均值。是最近邻方法的推广。knn 可以看作是kenerl 中用比较data dependent 的metrics的一种。svm 经常搭配kernel 使用。值得注意的是kernel和local regression之间存在着对立的关系
- base 函数，把函数看成积木块来构建更加复杂的函数，多项式是其中一种。我个人对这类函数感悟不多，希望之后能补足。

## 其他零碎知识拾遗：
1. (P12) If not include the intercept, the linear regression model is an affine set cutting the Y-axis at (0, intercept)

这里的affine set/function指的是仿射变换之后得到的，即变换的时候原点也会变换，由于原点的变换，affine transformation不是一种线性变换。知乎上比较好的解释：https://www.zhihu.com/question/20666664

2. (P12) the gradient function =  是该方程所有的偏分的集合

3. (P12) why does the gradient point to the steepest uphill direction? 
the gradient function 实际表示了一个变量变化对function本身变化的影响（在控制其他变量不变的情况下）
这里注意一个概念就是如果想要达到最大的变化，我们实际上就应该沿着这个偏分指出的方向走， 让变化最大化（三角形不等定理可以证明）
比较好的教程https://www.khanacademy.org/math/multivariable-calculus/multivariable-derivatives#partial-derivative-and-gradient-articles
这个教程中包含了一道题目来帮忙理解，个人觉得还是很到位的https://www.khanacademy.org/math/multivariable-calculus/multivariable-derivatives/partial-derivative-and-gradient-articles/a/directional-derivatives-going-deeper

4. 在豆瓣上阅读了大刀的笔记，其中有句话让我思考了很久 **“损失函数实际上是某种条件分布下的期望”** ，个人觉得一半一半，首先从概念上讲，fit 一个model实际上就是在学习一个条件分布（bayes 思想），那么randomness 排除在外，我们本质上是希望能解释除了random外的部分，即这个条件分布下的期望要趋紧于0， 但有没有例外呢？ 有，当损失函数不是在度量预测和真值之前的差距的时候，就不符合这一定义了，比如说hinge loss；还有就是当损失函数不是通过整个data 来计算误差的时候，比如stochasitc gradient decent 和 batch gradient descent就是只通过一部分data 或者一个sample来计算loss，这时的loss function就不是基于整个data的了
